Close add HDR tone mapping settings, #4358

low-batt opened the pull request:
This commit will:
- Add the new settings `enableToneMapping`, `toneMappingTargetPeak` and `toneMappingAlgorithm`
- Update the `Audio/Video` settings panel to allow the user to control these settings
- Update `VideoView.refreshEdrMode` to support these new settings
- Add the `CoreDisplay` framework to the Xcode project
- Add support to `VideoView.refreshEdrMode` to automatically determine the target peak setting based on the peak brightness of the display using the `CoreDisplay` framework
- Add reporting of the pixel format to the Inspector window
- Enhance HDR detection in `InspectorWindowController.updateInfo`
- Correct the path to the PiP framework in the Xcode project

These change provide the ability to mpv's tone mapping when playing HDR videos.

The main author of these changes is Carter Li. This is a merge of changes that have been tested in his IINA fork. A few minor changes were made during the merge.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #4358.

---

**Description:**

low-batt committed:
Close add HDR tone mapping settings, #4358

This commit will:
- Add the new settings enableToneMapping, toneMappingTargetPeak and
  toneMappingAlgorithm
- Update the Audio/Video settings panel to allow the user to control
  these settings
- Update VideoView.refreshEdrMode to support these new settings
- Add the CoreDisplay framework to the Xcode project
- Add support to VideoView.refreshEdrMode to automatically determine
  the target peak setting based on the peak brightness of the display
  using the CoreDisplay framework
- Add reporting of the pixel format to the Inspector window
- Enhance HDR detection in InspectorWindowController.updateInfo
- Correct the path to the PiP framework in the Xcode project

These change provide the ability to mpv's tone mapping when playing HDR
videos.

The main author of these changes is Carter Li. This is a merge of
changes that have been tested in his IINA fork. A few minor changes
were made during the merge.

Co-authored-by: CarterLi <carter.li@eoitek.com>

low-batt commented:
### Notes for Reviewers

A few things of note…
- The changes add use of the CoreDisplay framework
- Display information returned by macOS is different in older versions
- The changes require the planned mpv upgrade as they will try and set `tone-mapping` to `auto` which is not available in mpv 0.34.1. Nothing drastically bad happens, you just need to change the setting to avoid mpv complaining.
- The `Target peak` setting only accepts a number, it does not accept `auto` as the mpv [target-peak](https://mpv.io/manual/stable/#options-target-peak) does

### Changes Made During Merge

I made some changes during the merge. Hopefully they seem reasonable, if not object!

#### Audio/Video Settings Panel

I indented the controls for `Target peak` and `Algorithm` as they are disabled when enable tone mapping is disabled. That makes them a part of a "hierarchy of settings". The HIG indicates indentation should be used to communicate that. Other similar IINA settings use indentation.

I greatly simplified the descriptions of the new settings to match the existing "minimal" style. Of note the original descriptions contain URLs pointing to the relevant section of the mpv manual:
```text
Maximum luminance supported by your screen
if set to 0, IINA will try to detect the maximum luminance of current monitor
Ref: https://mpv.io/manual/master/#options-target-peak
```

That is an interesting idea. Of course these should be clickable links and we'd have to discuss whether they should point to a version specific manual rather than master. Since this is not done for other settings I thought it best to consider such a change for the feature release and update the descriptions of many settings.

I this particular case referencing the mpv manual might make you think you could use the value `auto`, but as noted above that is not accepted.

I also tried to correct constraint warnings reported by Xcode. @svobs, If you have time please take a look and see if I'm adding regressions with respects to layout constraints.

#### Logging

To avoid the large log message that results from logging the full display info dictionary I changed the logging in `VideoView.requestEdrMode` to only log the dictionary if we are unable to find either of the two keys.

#### Preference Handling

I changed the code in `VideoView.requestEdrMode` that obtains the algorithm from the preference setting to fall back to the default value if the preference value could not be converted into a mpv string. This matches up with how other code uses the `HardwareDecoderOption` preference enumeration. This prevents a crash if a user does something like this:
```bash
defaults write com.colliderli.iina.plist toneMappingAlgorithm 99
```

In other words even though users should not be messing with the preference file behind IINA's back we treat it as external input and provide some protection against invalid values.

#### Screen Shots

With the changes in this PR:
![issue-4358-merged](https://user-images.githubusercontent.com/86170219/232657548-4a25c81c-268d-46da-8550-d8836a593f02.jpg)

The original code with detailed descriptions of the settings:
![issue-4358-orig](https://user-images.githubusercontent.com/86170219/232657594-1851c6ec-cd51-46fe-9a5f-96dac762a03c.jpg)

CarterLi commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
     mpv.setString(MPVOption.GPURendererOptions.targetPrim, primaries)
+    // PQ videos will be display as it was, HLG videos will be converted to PQ
+    mpv.setString(MPVOption.GPURendererOptions.targetTrc, "pq")
+    mpv.setFlag(MPVOption.Screenshot.screenshotTagColorspace, true)
```
This was used for screenshot in [jxl format](https://github.com/CarterLi/iina/blob/develop/iina/Preference.swift#L538).

Since [(newer) mpv actually doesn't support ppm / pgm ...](https://[mpv.io](https://mpv.io/manual/master/#options-screenshot-format)/manual/master/#options-screenshot-format), I suggest merging my screenshot code too.

low-batt replied:
I had not noticed the old screenshot formats. They were removed from the UI in commit 994d6228202221a8ffccc74c29aa3530ec5366f5 back in 2017. I guess they did not remove the values from the enum to preserve compatibility with old plists. Probably safe to remove them now. I found the IINA+ commit that adds the new formats. I will prepare a separate PR for that change. Thanks for pointing this out.

CarterLi commented:
I like the style change in preferences window. Good work!

However, as tone-mapping algorithm is hard to understand for most people, I still prefer keeping the link to MPV document.

low-batt commented:
I do like the idea of linking to the mpv manual. If the other reviewers approve I can make that change. I think they need to be clickable links. I'm thinking they should link to the stable manual and not master. Thoughts reviewers?

uiryuu commented:
@low-batt Can you rebase this PR? I'm having a hard time building this PR since we've already updated our deps.

We should align the new options with previous options. I understand you are trying to align the texts, but this seems off to me. The original design looks better to me in this sense.
![image](https://user-images.githubusercontent.com/20237141/232964011-3b1d85dd-932e-493f-8cc9-0dfa5835b823.png)

As for link to the mpv manual, we can definitely do that, but adding a clickable button to the link of mpv manual is better, just like we previously did:
<img width="208" alt="image" src="https://user-images.githubusercontent.com/20237141/232964237-1977deed-bc95-42f1-bcfe-cfd723c755bf.png">

low-batt commented:
I will rebase and fix the merge conflicts. I will add `?` links to mpv manual.

Late here. I will get this done tomorrow.

low-batt committed:
Close add HDR tone mapping settings, #4358

This commit will:
- Add the new settings enableToneMapping, toneMappingTargetPeak and
  toneMappingAlgorithm
- Update the Audio/Video settings panel to allow the user to control
  these settings
- Add help buttons that link to Wikipedia and the appropriate sections
  of the mpv manual
- Update VideoView.refreshEdrMode to support these new settings
- Add the CoreDisplay framework to the Xcode project
- Add support to VideoView.refreshEdrMode to automatically determine
  the target peak setting based on the peak brightness of the display
  using the CoreDisplay framework
- Add reporting of the pixel format to the Inspector window
- Enhance HDR detection in InspectorWindowController.updateInfo
- Correct the path to the PiP framework in the Xcode project

These change provide the ability to mpv's tone mapping when playing HDR
videos.

The main author of these changes is Carter Li. This is a merge of
changes that have been tested in his IINA fork. A few minor changes
were made during the merge.

Co-authored-by: CarterLi <carter.li@eoitek.com>

low-batt commented:
I rebased with `develop` and corrected the merge conflicts.

I added help buttons that link to:
- https://en.wikipedia.org/wiki/Tone_mapping
- https://mpv.io/manual/stable/#options-target-peak
- https://mpv.io/manual/stable/#options-tone-mapping

The panel now looks like:

![issue-4358-help](https://user-images.githubusercontent.com/86170219/233191445-19acd3ac-1b01-4f44-b2c3-cab0bf797b24.jpg)

uiryuu commented on iina/PrefCodecViewController.swift:
```diff
@@ -99,4 +103,21 @@ class PrefCodecViewController: PreferenceViewController, PreferenceWindowEmbedda
     hwdecDescriptionTextField.stringValue = hwdec.localizedDescription
   }
+
+  @IBAction func updateEnableToneMapping(_ sender: NSButton) {
```
Can you do this in xib by bindings?

low-batt replied:
Yes, I'm changing this to binding enabling of the two controls to the checkbox in the XIB.

uiryuu commented on iina/VideoView.swift:
```diff
@@ -310,12 +310,16 @@ class VideoView: NSView {
+      player.mpv.setString(MPVOption.GPURendererOptions.targetPeak, "auto")
+      player.mpv.setString(MPVOption.GPURendererOptions.toneMapping, "auto")
+      player.mpv.setString(MPVOption.GPURendererOptions.toneMappingParam, "default")
+      player.mpv.setString(MPVOption.GPURendererOptions.toneMapping, "")
```
Why did you set this option twice? On line 319
```
player.mpv.setString(MPVOption.GPURendererOptions.toneMapping, "auto")
```

low-batt replied:
Because it is that way in the code in the other repository and I failed to notice that when preparing the merge. GOOD CATCH!

The "auto" for toneMapping was a recent add by mpv, so maybe the empty string setting was supposed to be removed after upgrading to 035.1?

@CarterLi Which is the correct value to use?

I'm going to guess "auto" and remove the setting to the empty string.

CarterLi replied:
Just use the default value of MPV

low-batt replied:
The default is "auto", so I removed this [line](https://github.com/CarterLi/iina/blob/a7e5cb961b8f4b209d3532e8653ef73f0d88415c/iina/VideoView.swift#L292) that is in IINA+

uiryuu commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
+            targetPeak = hdrLuminance
+          } else if let hdrLuminance = displayInfo["DisplayBacklight"] as? Int {
+            // We know macOS Catalina uses this key.
+            Logger.log("Found DisplayBacklight: \(hdrLuminance)", subsystem: hdrSubsystem);
```
Remove ;

low-batt replied:
I did a search in this source and removed a number of semicolons.

uiryuu commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
+          Logger.log("Successfully obtained information about the display", subsystem: hdrSubsystem)
+          // Prefer ReferencePeakHDRLuminance, which is reported by newer macOS versions.
+          if let hdrLuminance = displayInfo["ReferencePeakHDRLuminance"] as? Int {
+            Logger.log("Found ReferencePeakHDRLuminance: \(hdrLuminance)", subsystem: hdrSubsystem);
```
Remove ;

uiryuu commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
+            Logger.log("Found DisplayBacklight: \(hdrLuminance)", subsystem: hdrSubsystem);
+            targetPeak = hdrLuminance
+          } else {
+            Logger.log("Didn't find ReferencePeakHDRLuminance or DisplayBacklight, assuming HDR400", subsystem: hdrSubsystem);
```
Remove ;

uiryuu commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
+            targetPeak = 400
+          }
+        } else {
+          Logger.log("Unable to obtain display information", level: .warning, subsystem: hdrSubsystem);
```
Remove ;

uiryuu commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
+      let algorithm = Preference.ToneMappingAlgorithmOption(rawValue: Preference.integer(for: .toneMappingAlgorithm))?.mpvString
+        ?? Preference.ToneMappingAlgorithmOption.defaultValue.mpvString
+
+      Logger.log("Will enable tone mapping target-peak=\(targetPeak) algorithm=\(algorithm)", subsystem: hdrSubsystem);
```
Remove ;

uiryuu commented on iina/VideoView.swift:
```diff
@@ -393,8 +397,45 @@ extension VideoView {
+            targetPeak = 400
+          }
+        } else {
+          Logger.log("Unable to obtain display information", level: .warning, subsystem: hdrSubsystem);
```
We should also fallback the target peak to 400 if the display info cannot be obtained. Otherwise if we cannot obtain  the display info, `MPVOption.GPURendererOptions.targetPeak` will be set to 0 later.

low-batt replied:
I've updated this to also default to 400.

uiryuu reviewed (changes requested):
Should we call `refreshEdrMode` when the display is changed? A quick search shows that we are not doing that currently, but I might have missed that.

From a UI point of view, I think the text field under "Enable tone mapping" and the text field under the algorithm drop down do not provide useful information, given that we already provide the help link to the users. If they are confused by the settings, these text fields would not help.

For the text field under "Target peak", should we also inform the users that if we cannot detect automatically, the nits value will fallback to 400 nits?

low-batt committed:
Close add HDR tone mapping settings, #4358

This commit will:
- Add the new settings enableToneMapping, toneMappingTargetPeak and
  toneMappingAlgorithm
- Update the Audio/Video settings panel to allow the user to control
  these settings
- Update VideoView.refreshEdrMode to support these new settings
- Add the CoreDisplay framework to the Xcode project
- Add support to VideoView.refreshEdrMode to automatically determine
  the target peak setting based on the peak brightness of the display
  using the CoreDisplay framework
- Add reporting of the pixel format to the Inspector window
- Enhance HDR detection in InspectorWindowController.updateInfo
- Correct the path to the PiP framework in the Xcode project

These change provide the ability to mpv's tone mapping when playing HDR
videos.

The main author of these changes is Carter Li. This is a merge of
changes that have been tested in his IINA fork. A few minor changes
were made during the merge.

Co-authored-by: CarterLi <carter.li@eoitek.com>

low-batt commented:
This sequence shows the window starting out on my external LG display which does not support HDR. Then moving to the internal XDR display which does support HDR. Then moving back. IINA enabled and disabled HDR appropriately:
<details>
<summary><b>IINA log:</b></summary>

```text
21:01:01.361 [iina][d] Refreshing HDR for player0 @ display3: "LG UltraFine" visible frame (0.0, 0.0, 2560.0, 1415.0) EDR: {supports=false maxPotential=1.0 maxCurrent=1.0}
21:01:01.361 [hdr][d] HDR gamma=pq, primaries=bt.2020, sig_peak=4.926108360290527
21:01:01.361 [hdr][d] HDR video was found but the display does not support EDR mode
21:01:01.361 [hdr][d] Loading ICC profile
21:01:05.189 [iina][d] Falling back to nominal display refresh rate: 120.0 from 59.9990747912984
21:01:05.189 [iina][d] Refreshing HDR for player0 @ display1: "Built-in Retina Display" visible frame (458.0, -1056.0, 1728.0, 1018.0) EDR: {supports=true maxPotential=16.0 maxCurrent=1.0}
21:01:05.189 [hdr][d] HDR gamma=pq, primaries=bt.2020, sig_peak=4.926108360290527
21:01:05.190 [hdr][d] Will activate HDR color space instead of using ICC profile
21:01:05.191 [hdr][d] Successfully obtained information about the display
21:01:05.191 [hdr][d] Found ReferencePeakHDRLuminance: 1000
21:01:05.191 [hdr][d] Will enable tone mapping target-peak=1000 algorithm=auto
21:01:21.269 [iina][d] Falling back to nominal display refresh rate: 59.99925000937488 from 120.25120635180502
21:01:21.270 [iina][d] Refreshing HDR for player0 @ display3: "LG UltraFine" visible frame (0.0, 0.0, 2560.0, 1415.0) EDR: {supports=false maxPotential=1.0 maxCurrent=1.0}
21:01:21.270 [hdr][d] HDR gamma=pq, primaries=bt.2020, sig_peak=4.926108360290527
21:01:21.270 [hdr][d] HDR video was found but the display does not support EDR mode
21:01:21.270 [hdr][d] Loading ICC profile
```
</details>

I removed the two unhelpful text fields and changed the remaining one to indicate the fall back to 400. The panel now looks like:
![issue-4358-review](https://user-images.githubusercontent.com/86170219/233814162-d25627df-cd2f-43fd-8b59-13da348fc469.jpg)

uiryuu committed:
Add a period
